### ISSUE TITLE
[BUG] Boorus not retaining since multi booru update

### ISSUE THREAD

**Issue by Data-PhiIe** [NO-ob/LoliSnatcher_Droid#105]
Boorus not retaining since multi-booru update (keep having to re-add them every time I start the app), I can now back up boorus unlike before, however, restoring from backed up boorus doesn't work either. I've been kind of patiently waiting for a fix since multi-booru was introduced but figure by this point I should submit an issue.

Device is a Pixel 3 with Graphene OS

**Comment by NO-ob** [comment]
not sure what would be causing this but there haven't been other reports it could be your rom. 

could you try making a logcat while adding some boorus. install adb on your pc, connect your phone to your pc, then type in terminal ```adb logcat > output.txt``` 

after that add some boorus and then post the file here

>I've been kind of patiently waiting for a fix since multi-booru was introduced but figure by this point I should submit an issue.
patiently waiting for a fix for an issue you didn't report? nice one

**Comment by Data-PhiIe** [comment]
> not sure what would be causing this but there haven't been other reports it could be your rom.
> 
> could you try making a logcat while adding some boorus. install adb on your pc, connect your phone to your pc, then type in terminal `adb logcat > output.txt`
> 
> after that add some boorus and then post the file here
> 
> > I've been kind of patiently waiting for a fix since multi-booru was introduced but figure by this point I should submit an issue.
> 
> patiently waiting for a fix for an issue you didn't report? nice one

Yeah I was banking on someone else having the issue tends to happen, didn't this time, not blaming anyone but myself for that one, lol... it just became enough of an inconvenience for me to bother, love the app BTW.
I'll make a log tomorrow afternoon, but in the meantime if it helps graphene OS in particular is a hardened android fork IDK if their hardening methods would cause any issues with the app in particular.

**Comment by Data-PhiIe** [comment]
> > not sure what would be causing this but there haven't been other reports it could be your rom.
> > could you try making a logcat while adding some boorus. install adb on your pc, connect your phone to your pc, then type in terminal `adb logcat > output.txt`
> > after that add some boorus and then post the file here
> > > I've been kind of patiently waiting for a fix since multi-booru was introduced but figure by this point I should submit an issue.
> > 
> > 
> > patiently waiting for a fix for an issue you didn't report? nice one
> 
> Yeah I was banking on someone else having the issue tends to happen, didn't this time, not blaming anyone but myself for that one, lol... it just became enough of an inconvenience for me to bother, love the app BTW. I'll make a log tomorrow afternoon, but in the meantime if it helps graphene OS in particular is a hardened android fork IDK if their hardening methods would cause any issues with the app in particular.

**Comment by Data-PhiIe** [comment]
I went ahead and filtered everything relevant to the app from the log, hope it helps! :)

Line 2195: 03-17 05:40:15.587  1437  1722 I SystemServerTimingAsync: InitThreadPoolExec:Update app-ops uidState in case package com.noaisu.loliSnatcher changed
	Line 2843: 03-17 05:40:16.089  1437  1437 V StorageManagerService: Package com.noaisu.loliSnatcher does not have legacy storage
	Line 5650: 03-17 05:53:55.531  1437  1467 I ActivityTaskManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=com.noaisu.loliSnatcher/.MainActivity bnds=[816,1021][1020,1247]} from uid 10127
	Line 5664: 03-17 05:53:55.633  1437  1612 I ActivityManager: Start proc 6565:com.noaisu.loliSnatcher/u0a321 for pre-top-activity {com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 5664: 03-17 05:53:55.633  1437  1612 I ActivityManager: Start proc 6565:com.noaisu.loliSnatcher/u0a321 for pre-top-activity {com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 5664: 03-17 05:53:55.633  1437  1612 I ActivityManager: Start proc 6565:com.noaisu.loliSnatcher/u0a321 for pre-top-activity {com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 5724: 03-17 05:54:19.011  1437  4079 I ClipboardService: LoliSnatcher pasted from your clipboard
	Line 5787: 03-17 05:54:43.232  1437  1929 E ClipboardService: Denying clipboard access to com.noaisu.loliSnatcher, application is not in focus nor is it a system service for user 0
	Line 5791: 03-17 05:54:52.388  1437  6553 I ClipboardService: LoliSnatcher pasted from your clipboard
	Line 5843: 03-17 05:55:48.915  1437  1603 I ActivityManager: Killing 6565:com.noaisu.loliSnatcher/u0a321 (adj 900): remove task
	Line 5844: 03-17 05:55:48.979  1437  1907 I WindowManager: WIN DEATH: Window{817df2f u0 com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 5844: 03-17 05:55:48.979  1437  1907 I WindowManager: WIN DEATH: Window{817df2f u0 com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 5849: 03-17 05:55:48.985  1437  1610 W UsageStatsService: Unexpected activity event reported! (com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity event : 23 instanceId : 243335520)
	Line 5849: 03-17 05:55:48.985  1437  1610 W UsageStatsService: Unexpected activity event reported! (com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity event : 23 instanceId : 243335520)
	Line 5861: 03-17 05:55:53.681  1437  1467 I ActivityTaskManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=com.noaisu.loliSnatcher/.MainActivity bnds=[816,1021][1020,1247]} from uid 10127
	Line 5865: 03-17 05:55:53.731  1437  1612 I ActivityManager: Start proc 7181:com.noaisu.loliSnatcher/u0a321 for pre-top-activity {com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 5865: 03-17 05:55:53.731  1437  1612 I ActivityManager: Start proc 7181:com.noaisu.loliSnatcher/u0a321 for pre-top-activity {com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 5865: 03-17 05:55:53.731  1437  1612 I ActivityManager: Start proc 7181:com.noaisu.loliSnatcher/u0a321 for pre-top-activity {com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 5871: 03-17 05:56:00.218  1437  1603 I ActivityManager: Killing 7181:com.noaisu.loliSnatcher/u0a321 (adj 905): remove task
	Line 5872: 03-17 05:56:00.280  1437  1929 I WindowManager: WIN DEATH: Window{38e7083 u0 com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 5872: 03-17 05:56:00.280  1437  1929 I WindowManager: WIN DEATH: Window{38e7083 u0 com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 5877: 03-17 05:56:00.285  1437  1610 W UsageStatsService: Unexpected activity event reported! (com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity event : 23 instanceId : 146899742)
	Line 5877: 03-17 05:56:00.285  1437  1610 W UsageStatsService: Unexpected activity event reported! (com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity event : 23 instanceId : 146899742)
	Line 6331: 03-17 06:01:56.225  1437  6553 I ActivityTaskManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=com.noaisu.loliSnatcher/.MainActivity bnds=[816,1021][1020,1247]} from uid 10127
	Line 6345: 03-17 06:01:56.327  1437  1612 I ActivityManager: Start proc 8151:com.noaisu.loliSnatcher/u0a321 for pre-top-activity {com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 6345: 03-17 06:01:56.327  1437  1612 I ActivityManager: Start proc 8151:com.noaisu.loliSnatcher/u0a321 for pre-top-activity {com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 6345: 03-17 06:01:56.327  1437  1612 I ActivityManager: Start proc 8151:com.noaisu.loliSnatcher/u0a321 for pre-top-activity {com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 6370: 03-17 06:02:49.608  1437  7177 I ClipboardService: LoliSnatcher pasted from your clipboard
	Line 6435: 03-17 06:03:15.646  1437  6553 E ClipboardService: Denying clipboard access to com.noaisu.loliSnatcher, application is not in focus nor is it a system service for user 0
	Line 6436: 03-17 06:03:15.648  1437  6553 E ClipboardService: Denying clipboard access to com.noaisu.loliSnatcher, application is not in focus nor is it a system service for user 0
	Line 6437: 03-17 06:03:15.648  1437  6553 E ClipboardService: Denying clipboard access to com.noaisu.loliSnatcher, application is not in focus nor is it a system service for user 0
	Line 6438: 03-17 06:03:15.648  1437  6553 E ClipboardService: Denying clipboard access to com.noaisu.loliSnatcher, application is not in focus nor is it a system service for user 0
	Line 6439: 03-17 06:03:15.649  1437  6553 E ClipboardService: Denying clipboard access to com.noaisu.loliSnatcher, application is not in focus nor is it a system service for user 0
	Line 6440: 03-17 06:03:15.649  1437  6553 E ClipboardService: Denying clipboard access to com.noaisu.loliSnatcher, application is not in focus nor is it a system service for user 0
	Line 6444: 03-17 06:03:28.038  1437  7177 I ClipboardService: LoliSnatcher pasted from your clipboard
	Line 6463: 03-17 06:04:19.515  1437  1603 I ActivityManager: Killing 8151:com.noaisu.loliSnatcher/u0a321 (adj 905): remove task
	Line 6464: 03-17 06:04:19.590  1437  5611 I WindowManager: WIN DEATH: Window{66fec49 u0 com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 6464: 03-17 06:04:19.590  1437  5611 I WindowManager: WIN DEATH: Window{66fec49 u0 com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 6469: 03-17 06:04:19.596  1437  1610 W UsageStatsService: Unexpected activity event reported! (com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity event : 23 instanceId : 209950492)
	Line 6469: 03-17 06:04:19.596  1437  1610 W UsageStatsService: Unexpected activity event reported! (com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity event : 23 instanceId : 209950492)
	Line 6482: 03-17 06:04:24.784  1437  5728 I ActivityTaskManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=com.noaisu.loliSnatcher/.MainActivity bnds=[816,1021][1020,1247]} from uid 10127
	Line 6486: 03-17 06:04:24.842  1437  1612 I ActivityManager: Start proc 8591:com.noaisu.loliSnatcher/u0a321 for pre-top-activity {com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 6486: 03-17 06:04:24.842  1437  1612 I ActivityManager: Start proc 8591:com.noaisu.loliSnatcher/u0a321 for pre-top-activity {com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 6486: 03-17 06:04:24.842  1437  1612 I ActivityManager: Start proc 8591:com.noaisu.loliSnatcher/u0a321 for pre-top-activity {com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 19077: 03-17 06:19:32.682  1437  1603 I ActivityManager: Killing 8591:com.noaisu.loliSnatcher/u0a321 (adj 900): remove task
	Line 19078: 03-17 06:19:32.827  1437  1710 I WindowManager: WIN DEATH: Window{6e40523 u0 com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 19078: 03-17 06:19:32.827  1437  1710 I WindowManager: WIN DEATH: Window{6e40523 u0 com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity}
	Line 19079: 03-17 06:19:32.828  1437  1710 W InputManager-JNI: Input channel object '6e40523 com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity (client)' was disposed without first being removed with the input manager!
	Line 19079: 03-17 06:19:32.828  1437  1710 W InputManager-JNI: Input channel object '6e40523 com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity (client)' was disposed without first being removed with the input manager!
	Line 19086: 03-17 06:19:32.838  1437  1610 W UsageStatsService: Unexpected activity event reported! (com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity event : 23 instanceId : 35653694)
	Line 19086: 03-17 06:19:32.838  1437  1610 W UsageStatsService: Unexpected activity event reported! (com.noaisu.loliSnatcher/com.noaisu.loliSnatcher.MainActivity event : 23 instanceId : 35653694)

**Comment by NO-ob** [comment]
Did you paste stuff there those clipboard logs look strange. Also i dont see anything about it failing to write booru configs did you add any?

**Comment by Data-PhiIe** [comment]
I just pulled the log into Notepad++ and filtered LoliSnatcher, from there I copy pasta'd all relevant logs to that key word here (there was a lot of irrelevant noise in the log). Is there any other keywords I should be filtering for you?

**Comment by Data-PhiIe** [comment]
> 
I did paste the URL for the booru form my clipboard

**Comment by NO-ob** [comment]
okay good was thinking we coulda  had a pozzed dependency, i think it would show as lolisnatcher in the logs if it was failing to write. I do need a new phone so if you wait like 2 weeks ill get a pixel and install graphene os on there this hasn't been reported before so I assume its due to the rom

**Comment by Data-PhiIe** [comment]
Thank you, I appreciate it! :) I like grapheneOS as far as de-googled options are concerned, It's be a good ROM! I actually just ordered a Pixel 5a since my pixel three has fallen into legacy support this year, so I'll let you know If I have any issues on that device as well so we can see if its ROM or device specific.

**Comment by Data-PhiIe** [comment]
I just wanted to give an update, It's working fine on my 5a, it could just be an issue with the pixel 3 (since its in legacy support now). or it could be that I need to clean install the app again, but since I'm moving everything over to the 5a anyways I'm just not going to worry about it on my end. However if you want me to help troubleshoot it further on my end to see about fixing it for other potential Pixel users, just go ahead and let me know! :)

**Comment by vsr3y** [comment]
I've had this issue on my Samsung S20 Ultra despite having the latest version (V. 2.2.5+171 and I've isolated the cause. It seems that using some illegal characters to name boorus during "Add new Booru" and "Edited selected (Booru)" will cause LoliSnatcher to forget those boorus upon relaunch. 
What causes the problem for me is using two or more forward slashes with at least one character somewhere before the first slash or between slashes; there is no issue in the case of Booru names with characters existing only after the last slash.

For instance, Booru names that cause the issue are:
"GF//"
"/GF/"
"Gravity/Falls/"
"/Gravity/Falls"
"Gra/vity/Falls"
"Gra/vi/ty/Falls"

In contrast, Booru names that won't cause the issue:
"//"
"//GF"
"///GF"

I think it's a parsing issue.

**Comment by NO-ob** [comment]
Adding a / will make it try and write to a directory that doesn't exist

**Comment by NANI-SORE** [comment]
In 2.3 booru names are now sanitized to avoid breaking file paths

**Comment by vsr3y** [comment]
There's a closely-related and mostly benign issue to this one so I'm putting it here instead of opening a new issue.

Adding dots to boorus' names with make LoliSnatcher download images with only characters from the start of the name up until the first dot encountered, omitting the subsequent underscore and MD5 hash, e.g.
- Booru name ".e621" yields downloads:
> ".jpg"
> " (1).jpg"
> " (2).jpg"
> ...
- Booru name "e6.21" yields downloads:
> "e6.jpg"
> "e6 (1).jpg"
> "e6 (2).jpg"
> ...

**Comment by NANI-SORE** [comment]
Yeah, current logic just splits file name by dots to separate file extension and takes only the first part of that split.
It didn't even come to my mind that someone could use dots in booru names, I'll change it to take every part from split except the last one.
(But for now I'd recommend to remove dots from your booru names)

P.S. 😭 was there from the start 😜

**Comment by vsr3y** [comment]
Yea, since there are so many Rule 34 websites, I would just name boorus ".paheal", ".xxx", ".world", and stuff like that.
